### DISCUSSION TITLE
Help needed with decoding H264 into frames

### DISCUSSION THREAD

**Discussion by Valkhes** [asticode/go-astiav#62] [General]
Hello, I'm asking here because I can't figure what I'm supposed to do with the library. 

I'm trying to change a code I have in python using PyAV (which is a Python binding of FFMPEG) to Go, using go-astiav, as it seems to be one of the only maintained binding of FFMPEG in go.

I'm trying to retrieve H264 stream from a socket sended by a scrcpy-server (I know that some people have been using the lib here for the same reason). I read these one : 

https://github.com/asticode/go-astiav/issues/45
https://github.com/asticode/go-astiav/issues/13

But I don't exactly understand how I need to use go-astiav. I'm quite a beginner with H264 and codec, so this is a bit hard for me to understand.

In python, what I'm doing is : 

```
#Receive from the socket byte stream
raw_h264 = self.video_socket.recv(0x10000)

#Split up a byte stream into list of Packet
all_packets = self.codec.parse(raw_h264)

packets_index = 0
decoded_frame = None

while not decoded_frame:
    #Iterating over Packet to find a frame
    decoded_frame = codec.decode(all_packets[packets_index])
    packets_index += 1

if decoded_frame:
    #do something with frame decoded
```

In go, what I've tried to do for the moment : 

```
  fmt.Println("----> Creating stream")
  s := &stream{inputStream: nil}

  fmt.Println("----> Find IDH264 decoder")
  #Find decoder for H264
  if s.decCodec = astiav.FindDecoder(astiav.CodecIDH264); s.decCodec == nil {
    fmt.Println("main: codec is nil")
    return
  }

  // Alloc codec context
  if s.decCodecContext = astiav.AllocCodecContext(s.decCodec); s.decCodecContext == nil {
    fmt.Println("main: codec context is nil")
    return
  }
  defer s.decCodecContext.Free()

  // Open codec context
  if err := s.decCodecContext.Open(s.decCodec, nil); err != nil {
    fmt.Println("main: opening codec context failed: %w", err)
    return
  }

  // Alloc packet
  pkt := astiav.AllocPacket()
  defer pkt.Free()

  // Alloc frame
  f := astiav.AllocFrame()
  defer f.Free()

  //get bytes stream from socket

  buf := make([]byte, 0x10000)
  _, err = videoSocket.Read(buf)
  if err != nil {
    fmt.Println("unable to read from socket")
  }
  fmt.Println("Got H264 byte stream data")

  //Transform into Packet
  if err := pkt.FromData(buf); err != nil {
    fmt.Println(fmt.Errorf("main: packet header load failed: %w", err))
    return
  }
  fmt.Println("pkt size is : %v", pkt.Size())

  //What to do know ? How to decode ?
```

At this step, I'm pretty sure I managed to parse the byte stream to a Packet list (but I'm not really sure, I would need confirmation), but I don't understand what I'm supposed to do after if I want to find a Frame by decoding the packets. The naming are different from PyAV here, so I'm a bit lost. I was expecting to find a decode() function from the codec, as it knows it's H264 packets, but it seems it's not the case.

Can everybody help me about this ? I'm sure I'm not that far...

**Comment by Valkhes** [comment]
I saw this code : 

```
// Loop through packets
	for {
		// Read frame
		if err := inputFormatContext.ReadFrame(pkt); err != nil {
			if errors.Is(err, astiav.ErrEof) {
				break
			}
			log.Fatal(fmt.Errorf("main: reading frame failed: %w", err))
		}

		// Get stream
		s, ok := streams[pkt.StreamIndex()]
		if !ok {
			continue
		}

		// Send packet
		if err := s.decCodecContext.SendPacket(pkt); err != nil {
			log.Fatal(fmt.Errorf("main: sending packet failed: %w", err))
		}

		// Loop
		for {
			// Receive frame
			if err := s.decCodecContext.ReceiveFrame(f); err != nil {
				if errors.Is(err, astiav.ErrEof) || errors.Is(err, astiav.ErrEagain) {
					break
				}
				log.Fatal(fmt.Errorf("main: receiving frame failed: %w", err))
			}

			// Do something with decoded frame
			log.Printf("new frame: stream %d - pts: %d", pkt.StreamIndex(), f.Pts())
		}
	}
```

Which seems to be something I need to use, it really looks like it's iterating over the packets to find a frame.
I don't understand inputFormatContext here, how to replace it, as it seems it's not what I need.

```
// Alloc input format context
	inputFormatContext := astiav.AllocFormatContext()
	if inputFormatContext == nil {
		log.Fatal(errors.New("main: input format context is nil"))
	}
	defer inputFormatContext.Free()

	// Open input
	if err := inputFormatContext.OpenInput(*input, nil, nil); err != nil {
		log.Fatal(fmt.Errorf("main: opening input failed: %w", err))
	}
	defer inputFormatContext.CloseInput()

	// Find stream info
	if err := inputFormatContext.FindStreamInfo(nil); err != nil {
		log.Fatal(fmt.Errorf("main: finding stream info failed: %w", err))
	}
```

The inputFormatContext is taking an input which seems to be a string, so this is not really what I need I suppose.

**Comment by asticode** [comment]
@Valkhes first off, thanks for using `go-astiav` ❤️ 

The first thing I'd recommend (maybe you've done it already) is checking out the examples especially the [Demuxing/Decoding](https://github.com/asticode/go-astiav/blob/master/examples/demuxing_decoding/main.go) one that seems to fit your needs.

Before digging deeper in your problem, could you tell me whether you can receive the socket stream only via `videoSocket.Read(buf)`?

**Reply by Valkhes** [reply]
Hello, yes, I have been reading this specific exemple, as said in my comment, this is where I saw that I wasn't really far and only needed to see if it was possible to do something to match the inputFormatContext from the exemple 👍 

This would be easier for me as I'm trying to keep in simple for people to read on my side, and we always has been using a socket to read. The idea is that scrcpy-server is a server running on Android phones, and using ADB forward, we can get a port from the phone where we are going to receive the byte stream data through a socket (there is also another socket specifically for sending input to the phone)

I saw from a comment in an issue that someone managed to do it using zmq to connect to the socket, then used cereal, but I had struggle with cereal as it seems we need to pre-compile something to use it. I was not a big fan of this solution as I would love to do it using a simple read, as I'm doing in Python. This would be the perfect solution for me, even if it adds a bit of code.

If there is anything else to consume the socket, I would be glad to use it if it keep things simple (I'm not working alone on the project and basically I need it to be readable and possible to modify by everybody)

By the way, thanks for this lib, go-astiav seems way easier to use than a lot of other things. I'm still a bit struggling because that's a world I don't really know, but I'm glad it exists!

**Reply by asticode** [reply]
Which protocol is the socket using (UDP, RTP, TCP, ...)? And is the socket only transporting one video stream, or are there other streams as well in the socket? (or several sockets transporting each one stream)

**Reply by Valkhes** [reply]
The socket is using TCP, and I'm having only one video stream in the socket. So one socket, one stream here.

```
fmt.Println("Connecting video socket")
videoSocket, err := net.Dial("tcp", "127.0.0.1"+":"+strconv.Itoa(27183))
if err != nil {
  fmt.Println("Got error while connecting video socket:", err)
  return nil, err
}
```

The idea is that in the socket I will find a few bytes at the beginning giving some information about the phone, but at this point in my code I will only get H264 stream bytes as everything before has been consumed properly.

**Reply by asticode** [reply]
Can you know when h264 packets start/end in your socket stream or is it an information you don't have? (I'm asking all those questions to try to point you to the proper direction)

**Reply by Valkhes** [reply]
That's an information I don't have, everything was handled by PyAV after retrieving the data from the socket, in codec.parse(). It was parsing for me the byte into packets.

Parse code is here (in case it could help to understand what it's doing) : https://github.com/PyAV-Org/PyAV/blob/main/av/codec/context.pyx#L308


I'm not exactly sure how I would be able to know that, but I suppose that to transform into a packet with go-astiav, I would need to provide the bytes corresponding to a real packet to `pkt.FromData(buf)`

Does it mean that I need to parse my bytes to find for exemple an header that would tell me "this is where the packet start", so I can then get the full packet, and then find in this packet if there is a frame ?

Only thing I know at this point is the size of the phone image, as I know the width and the height of the phone. I suppose that I could get the size of a packet with this, if I know the image size and the header size maybe, but that's only guessing as I never tried to do that before.

in PyAV, it seems it's using av_parser_parse2 under codec.parse
https://stackoverflow.com/a/51087447 -> looks like this is where it's doing the job for me

**Reply by asticode** [reply]
I wouldn't parse manually the socket bytes into packets if I were you.

I would use a `FormatContext` with a custom `IOContext` which I'd feed the reads from the socket instead. The `FormatContext` would then provide proper packets which I'd then be able to decode to access the frame's data.

Therefore what you need to do:
- alloc a format context
- alloc an io context and provide a read callback that will read from the socket
- update the format context with the io context using [SetPb](https://github.com/asticode/go-astiav/blob/master/format_context.go#L128)
- when using `.OpenInput()`, you don't need to provide a url since you're using a custom io, but you need to provide the `h264` input format (second parameter of the function) since you know you have a raw h264 stream as input
- at that point, I don't know whether the format context will manage to probe the stream properly, but let's assume it will
- then you can use the rest of the example code you've copied to decode the frames

Let me know how it goes 👍 

PS: beware in the first code you posted `_, err = videoSocket.Read(buf)` you need to get the first returned value `n` as well since that's the number of bytes really read. Therefore the read buffer is `b[:n]` not `b`

**Reply by Valkhes** [reply]
Thanks a lot for the help, I'm going to try this as soon as possible and give feedbacks. Pretty sure it could be helping other people as it seems it's the only active binding of ffmpeg here, and a lot of people are trying to use scrcpy server with a go client!

**Reply by asticode** [reply]
❤️ 

I'll try to add an example for demuxing with a custom io 👍

**Reply by Valkhes** [reply]
That would be helpful, I'm trying to understand right now how to use the IOContext as there is no documentation about it.

From what I see, here is the signature of AllocIOContext

```
func astiav.AllocIOContext(bufferSize int, readFunc astiav.IOContextReadFunc, seekFunc astiav.IOContextSeekFunc, writeFunc astiav.IOContextWriteFunc) (ic *astiav.IOContext, err error)
```

It take a readFunc, which in my case I suppose would be something like : 

```
//Create the socket before calling readFromSocket
var videoSocket *net.TCPConn


func readFromSocket(b []byte) (n int, err error) {
	buf := make([]byte, 0x10000)
	n, err = videoSocket.Read(buf)
	if err != nil {
		fmt.Println("unable to read from socket")
		return 0, err
	}
	fmt.Println("Got H264 byte stream data")
	copy(b, buf[:n])
	return n, nil
}
```

the readFunc is the function that is going to read from my socket, from what I understand from your message. That mean that my videoSocket needs to be a global variable so I can access it from the function without passing it as an arg.
But then, I don't understand what I need to put for the first argument of AllocIOContext. The bufferSize is known during the read, so I may be misinterpreting something here.


Right now, I have something like that : 

```
         inputFormatContext := astiav.AllocFormatContext()
	defer inputFormatContext.Free()

        //Not sure what to do at this point

	var readFunc astiav.IOContextReadFunc
	readFunc = readFromSocket

	ioContext, _ := astiav.AllocIOContext(0, readFunc, nil, nil)
	defer ioContext.Free()

	inputFormatContext.SetPb(ioContext)

	// Open input
	if err := inputFormatContext.OpenInput("", inputFormatContext.InputFormat(), nil); err != nil {
		fmt.Println("main: opening input failed: %w", err)
	}
```

Thanks a lot for all the help again, I feel that I'm not really far from having something!

**Reply by asticode** [reply]
`bufferSize` (the first argument of AllocIOContext) is the maximum number of bytes that will be read at each read. It seems you've decided it should be `0x10000`. Your read func could be as simple as 
```go
func readFromSocket(b []byte) (n int, err error) {
	n, err = videoSocket.Read(b)
        if errors.Is(err, io.EOF) {
            return 0, astiav.ErrEof
        }
        return
}
```
which would result in
```go
astiav.AllocIOContext(0x10000, func(b []byte) (n int, err error) {
	n, err = videoSocket.Read(b)
        if errors.Is(err, io.EOF) {
            return 0, astiav.ErrEof
        }
        return
}, nil, nil)	
```


PS : you should not ignore the error in `ioContext, _ := astiav.AllocIOContext(0, readFunc, nil, nil)`, and return when there's an error

**Reply by Valkhes** [reply]
Alright that's it, now I can set the Pb!
(About the error, yes, it's of course something I'm adding, it still a bit in building as I'm trying for the moment to make something that work to get at least a frame, but I'll be sure to not miss these one. Coming from python and the try/catch, it's a bad habit I have)

```

        fmt.Println("----> AllocIOContext")
	ioContext, err := astiav.AllocIOContext(0x10000, func(b []byte) (n int, err error) {
		n, err = videoSocket.Read(b)
		if errors.Is(err, io.EOF) {
			return 0, astiav.ErrEof
		}
		return
	}, nil, nil)
	if err != nil {
		fmt.Println(fmt.Printf("error during AllocIOContext : %s", err.Error()))
		return
	}
	defer ioContext.Free()

	fmt.Println("----> SetPb")
	inputFormatContext.SetPb(ioContext)

	inputFormat := astiav.FindInputFormat("h264")
        if inputFormat == nil {
		fmt.Println("inputFormat is nil")
		return
	}
	fmt.Println(fmt.Printf("name : %s", inputFormat.Name()))

	// Open input
	fmt.Println("----> OpenInput")
	if err := inputFormatContext.OpenInput("", inputFormat, nil); err != nil {
		fmt.Println("main: opening input failed: %w", err)
                 return
	}
	defer inputFormatContext.CloseInput()
```

It seems that I don't crash anymore, I guess everything is open now, just need to continue 👍 

I will come back as soon as I have something working. Thanks again for everything!

EDIT : Seems it's not opened in fact, it's hanging inside ```OpenInput``` as it doesn't continue in my code. I may have still an issue.

**Reply by asticode** [reply]
> Seems it's not opened in fact, it's hanging inside OpenInput as it doesn't continue in my code. I may have still an issue.

Yeah I'm having the same problem when writing the example. I've pinpointed the problem but I need to do more tests. I will let you know

**Reply by asticode** [reply]
OK so on master (and tag `v0.15.0`) signature of `AllocIOContext` has changed: there's an additional `writable bool` parameter. You need to set it to `false` in your case and that `.OpenInput()` shouldn't block anymore.

Also, I've added a [custom IO demuxing example](https://github.com/asticode/go-astiav/blob/master/examples/custom_io_demuxing/main.go) which I hope will be useful 👍

**Reply by asticode** [reply]
Also you may be able to use `videoSocket.Read` directly as the `readFunc` of `AllocIOContext` which would simplify your code

**Reply by Valkhes** [reply]
You are a life savior. 
As soon as I updated, I started to receive my frames : 

```
2024/05/16 17:25:35 ffmpeg log: nal_unit_type: 1(Coded slice of a non-IDR picture), nal_ref_idc: 3 - class: h264 [AVCodecContext] @ 0x158e051b0 - level: 48
2024/05/16 17:28:29 new frame: stream 0 - pts: -9223372036854775808
```

I can see the log appears each time I move the screen of my phone. There is a little bit to do, but from now I'll be able to do it by myself I think.

Coming back I soon as I finished everything, in case I still have a question, but you helped me so much already!

**Reply by asticode** [reply]
❤️

**Reply by Valkhes** [reply]
Hello Asticode ! 

I have some feedbacks to provide now that I managed to retrieve frames. 

After retrieving frames, I tried to get the Image from it, so I could save it in a file and check if it was working correctly.

Here is the code I'm having just after retrieving the Frame : 

```
                 log.Printf("new frame: stream %d - pts: %d", pkt.StreamIndex(), f.Pts())
		fmt.Println("height :", f.Height())
		fmt.Println("width :", f.Width())
		fmt.Println("<---- got an", f.PictureType().String(), "frame with pixelFormat", f.PixelFormat().Name())
		image, err := f.Data().GuessImageFormat()
		if err != nil {
			fmt.Println("cannot guess image format :", err.Error())
			return
		}
```

Here is the print I'm having when trying to run my program : 

```
2024/05/17 14:53:46 new frame: stream 0 - pts: -9223372036854775808
height : 2216
width : 1080
<---- got an I frame with pixelFormat yuv420p

```

First thing I see is that I retrieve an height and a width which seems to be pretty close from what I need.
My android phone size is 1080x2220
It looks like I'm missing 4 lines here, which could be why I'm having issue after, but overall it seems I'm having something close.

When I print the pictureType and pixelFormat, I see that I'm having an Iframe (meaning for H264 that it's supposed to be a full frame from what I understand), in YUV420P. It's good, because it mean I'm supposed to be able to use ```f.Data().GuessImageFormat()``` as yuv420p is being checked inside.

From there, it is hanging in GuessImageFormat(), and I'm not really sure about what is happening. I was thinking it may be because of the 4 line that appears to be missing in my image size, or maybe there is something else going wrong here too.

That was it for my feedback, I'm going to dig more in order to find if I can get more information

**Reply by Valkhes** [reply]
inputFormatContext.FindStreamInfo(nil) is returning me these logs, not sure if it could be helping but just in case : 
```
2024/05/17 15:06:09 ffmpeg log: Before avformat_find_stream_info() pos: 0 bytes read:40457 seeks:0 nb_streams:1 - class: h264 [AVFormatContext] @ 0x143614630 - level: 48
2024/05/17 15:06:09 ffmpeg log: nal_unit_type: 7(SPS), nal_ref_idc: 3 - class: extract_extradata [AVBSFContext] @ 0x600000a44000 - level: 48
2024/05/17 15:06:09 ffmpeg log: nal_unit_type: 8(PPS), nal_ref_idc: 3 - class: extract_extradata [AVBSFContext] @ 0x600000a44000 - level: 48
2024/05/17 15:06:09 ffmpeg log: nal_unit_type: 5(IDR), nal_ref_idc: 3 - class: extract_extradata [AVBSFContext] @ 0x600000a44000 - level: 48
2024/05/17 15:06:09 ffmpeg log: nal_unit_type: 7(SPS), nal_ref_idc: 3 - class: h264 [AVCodecContext] @ 0x1344043c0 - level: 48
2024/05/17 15:06:09 ffmpeg log: nal_unit_type: 8(PPS), nal_ref_idc: 3 - class: h264 [AVCodecContext] @ 0x1344043c0 - level: 48
2024/05/17 15:06:09 ffmpeg log: nal_unit_type: 5(IDR), nal_ref_idc: 3 - class: h264 [AVCodecContext] @ 0x1344043c0 - level: 48
2024/05/17 15:06:09 ffmpeg log: Format yuv420p chosen by get_format(). - class: h264 [AVCodecContext] @ 0x1344043c0 - level: 48
2024/05/17 15:06:09 ffmpeg log: Reinit context to 1088x2224, pix_fmt: yuv420p - class: h264 [AVCodecContext] @ 0x1344043c0 - level: 40
2024/05/17 15:06:09 ffmpeg log: nal_unit_type: 1(Coded slice of a non-IDR picture), nal_ref_idc: 3 - class: h264 [AVCodecContext] @ 0x1344043c0 - level: 48
2024/05/17 15:06:09 ffmpeg log: nal_unit_type: 1(Coded slice of a non-IDR picture), nal_ref_idc: 3 - class: h264 [AVCodecContext] @ 0x1344043c0 - level: 48
2024/05/17 15:06:09 ffmpeg log: nal_unit_type: 1(Coded slice of a non-IDR picture), nal_ref_idc: 3 - class: h264 [AVCodecContext] @ 0x1344043c0 - level: 48
2024/05/17 15:06:09 ffmpeg log: nal_unit_type: 1(Coded slice of a non-IDR picture), nal_ref_idc: 3 - class: h264 [AVCodecContext] @ 0x1344043c0 - level: 48
2024/05/17 15:06:10 ffmpeg log: nal_unit_type: 1(Coded slice of a non-IDR picture), nal_ref_idc: 3 - class: h264 [AVCodecContext] @ 0x1344043c0 - level: 48
2024/05/17 15:06:10 ffmpeg log: nal_unit_type: 1(Coded slice of a non-IDR picture), nal_ref_idc: 3 - class: h264 [AVCodecContext] @ 0x1344043c0 - level: 48
```

**Reply by asticode** [reply]
When you say it's hanging at `GuessImageFormat()` you mean that it blocks there and never reaches after the execution of that function?

**Reply by Valkhes** [reply]
In fact, it doesn't hang, I misinterpreted what I had because my program wasn't finishing because of something else 

It seems I get my image.Image, then I try to save it : 

```
                 image, err := f.Data().GuessImageFormat()
		if err != nil {
			fmt.Println("cannot guess image format :", err.Error())
			return
		}
		file, err := os.Create("image.jpg")
		if err != nil {
			return
		}
		options := &jpeg.Options{Quality: 90} // Set JPEG quality, range is from 1 to 100 (higher means better quality)
		err = jpeg.Encode(file, image, options)
		if err != nil {
			fmt.Println("can not encode jpeg:", err.Error())
			return
		}
```

It save 591 bytes in my file, and the image is unreadable. I'm trying to check if I can see what I have in the image.Image, to see if it's valid or not (I suppose it isn't as I don't get the ful image with jpeg.Encode)

**Reply by asticode** [reply]
You need to use [.ToImage(image)](https://github.com/asticode/go-astiav/blob/master/frame_data.go#L139) to copy the content of the frame into your image

**Reply by Valkhes** [reply]
Ok, I see, I was thinking GuessImageFormat was going to do it for me as it was returning an image.
Do I need to specify something else than just the Data of the Frame ? Or do another operation before to convert something ?

When calling : 

```
var image image.Image
err = f.Data().ToImage(image)
if err != nil {
  fmt.Println("cannot use ToImage:", err.Error())
  return
}
```

I get this error : 

cannot use ToImage: astiav: image format is not handled

**Reply by Valkhes** [reply]
Got it, 

```
image := image.NewYCbCr(image.Rect(0, 0, width, height), image.YCbCrSubsampleRatio420)
```

it works ! 

thanks a lot for everything again ! 👍

**Reply by asticode** [reply]
I think I'll need to create an example for this part as well 🤔 

In fact you need to use both `.GuessImageFormat()` to guess the image format and then provide the returned image to `.ToImage()` to actually copy the content. That would be better than creating your own image.

```go
image, err := .Data().GuessImageFormat()
err = f.Data().ToImage(image)
```

**Reply by Valkhes** [reply]
Ah yes, tested it, that was it

Indeed that was implicit, GuessImageFormat is here only to define the image size/format and create it with the good ratio, and you need to fill it using ToImage after, it make sense.
Not sure if an exemple is necessary, it surely make sense for a lot of people, I'm just quite a beginner in this field so I still need to understand exactly what I need to do each time.

**Reply by asticode** [reply]
FYI I've added [an example on how to process frame data](https://github.com/asticode/go-astiav/commit/e0a9bdbd45d58486c3486495d09a40ce98d5a7fa) in the demuxing/decoding example 👍

**Reply by Valkhes** [reply]
Hello ! 

Everything is working like a charm with the exemple you gave, I managed to finish my decoding pipeline with everything we discussed.

I wanted to thank you again for everything!

I'm closing the discussion 👍

**Reply by asticode** [reply]
❤️
